### PR TITLE
Use Local Path Command Line Option

### PR DESCRIPTION
**Pull request by slackstone**
Directory Options:
--use-local-path

Sets the default path precedence for loading local game assets.

Add-on assets contained in the local SuperTux/addon home directory, that also match an expected game asset path, will take precedence.
 
This feature allows game hackers the ability to load their own game assets and add-on files.
 
 Example of addon home directory path, found on most Linux platforms: 
  ~/.local/share/supertux2/addons

### CODE REVIEW

File: src/addon/addon_manager.cpp [hunk 1]
```diff
@@ -445,18 +446,37 @@ AddonManager::enable_addon(const AddonId& addon_id)
         break;
     }
 
-    if (PHYSFS_mount(addon.get_install_filename().c_str(), mountpoint.c_str(), 1) == 0)
+    if (g_config->use_local_path)
     {
+      if (PHYSFS_mount(addon.get_install_filename().c_str(), mountpoint.c_str(), 0) == 0)
+      {
       log_warning << "Could not add " << addon.get_install_filename() << " to search path: "
                   << PHYSFS_getLastErrorCode() << std::endl;
```
Comment (Semphriss) [inline src/addon/addon_manager.cpp:454]: Add extra indent

File: src/addon/addon_manager.cpp [hunk 2]
```diff
@@ -445,18 +446,37 @@ AddonManager::enable_addon(const AddonId& addon_id)
...
+      else
+      {
+       if (addon.get_type() == Addon::LANGUAGEPACK)
+        {
+         PHYSFS_enumerate(addon.get_id().c_str(), add_to_dictionary_path, nullptr);
+         }
+       addon.set_enabled(true);
       }
-      addon.set_enabled(true);
     }
```
Comment (Semphriss) [inline src/addon/addon_manager.cpp:480]: Indentation problem, brackets lining is correct but make sure it's +2 spaces after an open bracket and -2 before a closed one

File: src/addon/addon_manager.cpp [hunk 3]
```diff
@@ -522,12 +542,25 @@ void
...
+        else
+        {
+        if (PHYSFS_mount(addon->get_install_filename().c_str(), mountpoint.c_str(), 1) == 0)
+        {
+                          log_warning << "Could not add " << addon->get_install_filename() << " to search path: "
+                          << PHYSFS_getLastErrorCode() << std::endl;
+        }
       }
     }
   }
```
Comment (Semphriss) [inline src/addon/addon_manager.cpp:566]: Indentation fix here to, there's an extra +2 tabs _before_ an open bracket after the `for` and the open bracket before the last `if` doesn't add +2 spaces. Also, some lines have way too many spaces :)

File: src/addon/addon_manager.cpp [hunk 4]
```diff
@@ -538,10 +571,11 @@ AddonManager::unmount_old_addons()
...
-      {
-        log_warning << "Could not remove " << addon->get_install_filename() << " from search path: "
-                    << PHYSFS_getLastErrorCode() << std::endl;
+      if (g_config->use_local_path)
+        if (PHYSFS_unmount(addon->get_install_filename().c_str()) == 0)
+        {
+          log_warning << "Could not remove " << addon->get_install_filename() << " from search path: "
+                     << PHYSFS_getLastErrorCode() << std::endl;
       }
     }
```
Comment (Semphriss) [inline src/addon/addon_manager.cpp:580]: Leave the wrong bracket lining on the lines you didn't write, but add brackets to `g_config->use_local_path` to make the code cleaner

File: src/addon/addon_manager.cpp [hunk 5]
```diff
@@ -645,7 +679,12 @@ AddonManager::add_installed_archive(const std::string& archive, const std::strin
   {
     std::string os_path = FileSystem::join(realdir, archive);
 
-    PHYSFS_mount(os_path.c_str(), nullptr, 1);
+ if (g_config->use_local_path) {
+    PHYSFS_mount(os_path.c_str(), nullptr, 0);
+    } else {
+ PHYSFS_mount(os_path.c_str(), nullptr, 1);
+    }
+
```
Comment (Semphriss) [inline src/addon/addon_manager.cpp:687]: `PHYSFS_mount(os_path.c_str(), nullptr, !(g_config->use_local_path));`

**Review by Semphriss — CHANGES_REQUESTED** [review]
Mostly just indent fixes. Thanks for the PR!

(I didn't test the code yet, Ima do that in a while but for the moment, if something I said seems to break the code, leave it there and lmk)

File: src/addon/addon_manager.cpp (hunk 4 above)
Comment (slackstone) [reply src/addon/addon_manager.cpp]: I've cleaned this up the best I can. I added brackets.

I think it's ready.

File: src/addon/addon_manager.cpp [hunk 6]
```diff
@@ -445,18 +446,37 @@ AddonManager::enable_addon(const AddonId& addon_id)
...
+    if (g_config->use_local_path)
     {
-      log_warning << "Could not add " << addon.get_install_filename() << " to search path: "
-                  << PHYSFS_getLastErrorCode() << std::endl;
+      if (PHYSFS_mount(addon.get_install_filename().c_str(), mountpoint.c_str(), 0) == 0)
+      {
+        log_warning << "Could not add " << addon.get_install_filename() << " to search path: "
+        << PHYSFS_getLastErrorCode() << std::endl;
+       }
+       else
```
Comment (Semphriss) [inline src/addon/addon_manager.cpp:456]: One space too much in front of those two lines

File: src/addon/addon_manager.cpp [hunk 7]
```diff
@@ -445,18 +446,37 @@ AddonManager::enable_addon(const AddonId& addon_id)
...
+      {
+        if (addon.get_type() == Addon::LANGUAGEPACK)
+          {
+             PHYSFS_enumerate(addon.get_id().c_str(), add_to_dictionary_path, nullptr);
+          }
```
Comment (Semphriss) [inline src/addon/addon_manager.cpp:461]: Two spaces too much in front of those three lines

File: src/addon/addon_manager.cpp [hunk 8]
```diff
@@ -445,18 +446,37 @@ AddonManager::enable_addon(const AddonId& addon_id)
...
+         log_warning << "Could not add " << addon.get_install_filename() << " to search path: "
+         << PHYSFS_getLastErrorCode() << std::endl;
+       }
+       else
+       {
+         if (addon.get_type() == Addon::LANGUAGEPACK)
+           {
+           PHYSFS_enumerate(addon.get_id().c_str(), add_to_dictionary_path, nullptr);
+           }
+         addon.set_enabled(true);
```
Comment (Semphriss) [inline src/addon/addon_manager.cpp:478]:
```
    {
      if (PHYSFS_mount(addon.get_install_filename().c_str(), mountpoint.c_str(), 1) == 0)
      {
        log_warning << "Could not add " << addon.get_install_filename() << " to search path: "
                    << PHYSFS_getLastErrorCode() << std::endl;
      }
      else
      {
        if (addon.get_type() == Addon::LANGUAGEPACK)
        {
          PHYSFS_enumerate(addon.get_id().c_str(), add_to_dictionary_path, nullptr);
        }
```

File: src/addon/addon_manager.cpp [hunk 9]
```diff
@@ -522,12 +542,25 @@ void
...
+            << PHYSFS_getLastErrorCode() << std::endl;
+          }
+       }
+       else
+       {
+         if (PHYSFS_mount(addon->get_install_filename().c_str(), mountpoint.c_str(), 1) == 0)
+           {
+             log_warning << "Could not add " << addon->get_install_filename() << " to search path: "
+             << PHYSFS_getLastErrorCode() << std::endl;
+        }
```
Comment (Semphriss) [inline src/addon/addon_manager.cpp:563]:
```
  for (auto& addon : m_installed_addons) 
  {
    if (is_old_enabled_addon(addon)) 
    {
      if (g_config->use_local_path) 
      {
        if (PHYSFS_mount(addon->get_install_filename().c_str(), mountpoint.c_str(), 0) == 0)
        {
          log_warning << "Could not add " << addon->get_install_filename() << " to search path: "
                      << PHYSFS_getLastErrorCode() << std::endl;
        }
      }
      else
      {
        if (PHYSFS_mount(addon->get_install_filename().c_str(), mountpoint.c_str(), 1) == 0)
        {
          log_warning << "Could not add " << addon->get_install_filename() << " to search path: "
                      << PHYSFS_getLastErrorCode() << std::endl;
        }
```

**Review by Semphriss — CHANGES_REQUESTED** [review]
Some more indentation problems, wrote the exact steps to fix or posted code snippets to make it easier

**Comment by Grumbel** [comment]
Is that option really necessary? Can't you just make that the default behavior and users that don't want to use it just don't copy addons there? There is already the option to enable/disable individual addons from within the game.

Either way, it would need a better name, `--use-local-path` really doesn't tell me much.

**Comment by slackstone** [comment]
This was the defult option up until about a year ago. Tobbi pushed an
update that removed this.

I first tried added this as a menu toggle and did a PR. People did not like.

The thing is, it's a potential support issue because a confused user might
think the game is broken if a addons does wacky suff with unexpected file
replacements. I saw this happening right around when Tobbi made the update,
i figured this was the prompt for the change.

It's hard to accidentally use a command line switch. This seemed like a
safe approach that would not upset or confuse people.

I would prefer just defaulting to this beghavior back to how it was. I use
this feature for my addon project.

I tried to catch Tobbi on the topic but did not push it.

It would be nice to have this feature back. I'm open to reanming the
switch, i tried really hard to come up with someting logical.

Thanks for the input.
Steve Groundwater / slackstone on github

**Pull request closed without merging**